### PR TITLE
Handle UDP responses that overflow with TC bit

### PR DESCRIPTION
**Pull request by SriHarsha001**
### 1. Why is this pull request needed and what does it do?
Description in this PR - https://github.com/coredns/coredns/pull/6003
Code changes in plugin/pkg/proxy/connect.go file is copied from above PR - https://github.com/coredns/coredns/pull/6003

This PR - https://github.com/coredns/coredns/pull/6100 was closed by mistake. So submitting a new PR here with the same code changes. 

Getting this PR https://github.com/miekg/dns/pull/1475 reviewed is taking time. Can we get the bug fix merged here with "overflow" string comparison. And we can address this "To accurately identify the error by type rather than string compares" as a follow up PR while I work with maintainers of miekg package. 

### 2. Which issues (if any) are related?
https://github.com/coredns/coredns/issues/5953
https://github.com/coredns/coredns/issues/5998

### 3. Which documentation changes (if any) need to be made?
Not sure. 

### 4. Does this introduce a backward incompatible change or deprecation?
No.

### CODE REVIEW

**Comment by SriHarsha001** [comment]
@chrisohaver - I apologize, I had accidently deleted my fork which closed this PR - https://github.com/coredns/coredns/pull/6100 , hence submitted this PR with the same code changes. 

As trying to get this PR https://github.com/miekg/dns/pull/1475 reviewed and merged is taking a long time, can we get the bug fix merged here with "overflow" string comparison.

Can we address this comment "To accurately identify the error by type rather than string compares" as a follow up PR while I work with maintainers of miekg package.

**Comment by SriHarsha001** [comment]
@chrisohaver - When you get a chance, I kindly request you to take a look at this request.

**Comment by chrisohaver** [comment]
Is there anything new to review here vs where https://github.com/coredns/coredns/pull/6100 was left off?

**Comment by SriHarsha001** [comment]
@chrisohaver - Thank you for your reply. I wanted to check if we get the bug fix merged here with "overflow" string comparison. And we can address this "To accurately identify the error by type rather than string compares" as a follow up PR while I work with maintainers of miekg package on this PR - https://github.com/miekg/dns/pull/1475

**Comment by chrisohaver** [comment]
> @chrisohaver - Thank you for your reply. I wanted to check if we get the bug fix merged here with "overflow" string comparison. And we can address this "To accurately identify the error by type rather than string compares" as a follow up PR while I work with maintainers of miekg package on this PR - [miekg/dns#1475](https://github.com/miekg/dns/pull/1475)

Aside from that, were there other outstanding issues that needed to be addressed?  It looks like there were ... relating to AA bit?  Maybe elsewhere?  Can you review the feedback and requested changes in https://github.com/coredns/coredns/pull/6100?

**Comment by SriHarsha001** [comment]
@chrisohaver - yes yes, we had discussed about setting AA bit to false here - https://github.com/coredns/coredns/pull/6100#discussion_r1242570014 

Please let me know, if there is anything outstanding that needs to be address.

**Comment by chrisohaver** [comment]
> I couldn't find any specific RFC that directly addresses the handling of the authorization bit (AA) in truncated DNS responses. There is no specific handling or interpretation defined for the authorization bit (AA) in truncated responses

I think in that case we should not alter the AA bit from the original.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Unless we can show that upstream implementations clear the AA and AD bits on a response they truncate _themselves_, I don't think we should alter the header bits from the original, with the exception of setting the TC bit

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Thank you. Sure, I have made the suggested changes.

File: plugin/pkg/proxy/proxy_test.go
Comment (chrisohaver) [inline plugin/pkg/proxy/proxy_test.go]: This isn't a performance concern since it's in a test, but to shorten code and make it more readable, please assign `ret.Answer` a single slice of tests, instead appending them one at a time.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: We only need to check this if the protocol is udp.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: I'm not sure how Go compiler optimizes and evals `||` expressions. Assuming it just checks left to right, it would be better to check `dnsErrBufOccured` before doing a the substring compare for "overflow".

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Why is this explicitly being copied over?  Is it not copied in `Copy()`?

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: IIUC, this must have already been set in `state.Req`.
So we need to explicitly set it here?

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: This case should be checked before making a copy of the response.  IIUC, it should be inverted and checked earlier.  Something like the following:

```
// drop response if its message id doesn't match the request message id.
if state.Req.Id != ret.Id {
  continue 
}
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, moved the protocol check. Once my PR for Miekg package is merged, I plan to update this code with below - 

			if proto == "udp" {
				switch err.(type) {
				case *dns.Error:
					if errors.Is(err, dns.ErrBuf) {
						dnsErrBufOccured = true
					}
				case *dns.OverflowError:
					dnsErrOverflowOccured = true
				default:
					dnsErrBufOccured = false
					dnsErrOverflowOccured = false
				}
			}

File: plugin/pkg/proxy/proxy_test.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/proxy_test.go]: Sure, made the suggested changes.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, made the suggested changes.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: IIUC - When a DNS server receives a query, it can respond with the RecursionAvailable bit set to indicate that it supports recursive resolution. So preserving RecursionAvailable field from the original response (ret).

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: The Response bit is set to indicate that message is a response sent by a DNS server to a client in answer to a query. The original request (state.Req.Copy()) will not have response field set. So preserving response field from the original response (ret).

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Moved this if condition up. Copying the original request state and setting TC bit only if the resp message ID matches request message ID.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [reply plugin/pkg/proxy/connect.go]: Yeah,  But isn't it already copied via `state.Req.Copy()`?

So setting it explicitly here is not necessary, right?

Comment (chrisohaver) [reply plugin/pkg/proxy/connect.go]: Ah, nevermind - I just realized this is copied from the _request_ - not the _response_...  which is counter intuitive.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Why is a new return message created by copying the _request_ (`state.Req`) instead of the response (`ret`)?

Can the code instead alter `ret` (i.e. set TC bit, and empty the answer/extra/auth sections?)

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Can `ret` be nil?

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: I was thinking about it on why the copy was made from the original request instead of original response. Not sure if there was any specific reason (which I don't know) why this was done by Patrick in his original PR. I will reach out to Patrick and see if I am missing something.

Comment (chrisohaver) [reply plugin/pkg/proxy/connect.go]: It looks like this was discussed in comments @phealy's original PR.  Original he used the `ret`, but now I see it changed, with the following rationale ...

> ... I think it would actually be cleaner to build a new response object that we guarantee is in a clean state.

I'm not convinced this is better though.  I think it would be simpler to use the original `ret`, set truncation bit, and clear out answer/extra/auth sections.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: yes, added a check in the if condition.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Yes, I have made the suggested changes and moved the logic in a new function. Would you mind taking another look please.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Is the copy necessary?  or can we modify the original response in place?

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Also need to clear out the Authority section (it's called Ns in the msg struct).

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: IMO, we should not modify this bit unless you have a compelling use case you can illustrate that this fixes.

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, modifying the original resp in place.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, clearing authority section.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, removed resetting AD bit here.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]:
```suggestion

```

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]:
```suggestion
			if proto == "udp" {
			// For UDP, if the error is an overflow, we probably have an upstream misbehaving in some way.
			// (e.g. sending >512 byte responses without an eDNS0 OPT RR).
			// Instead of returning an error, return an empty response with TC bit set. This will make the
			// client retry over TCP (if that's supported) or at least receive a clean
			// error. The connection is still good so we break before the close.

			isDNSBufferError := false
			isDNSOverflowError := false
			var perr *dns.Error

			
```

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]:
```suggestion
			
```

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go:133]:
```suggestion
}
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]
requested changes may be confusing...  I think all the error/overflow/truncation logic can be moved into the `if proto == "udp" { ... }` block.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: If `isDNSBufferError` is true, we don't need to do this substring string compare.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: where is `perr` used?

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [reply plugin/pkg/proxy/connect.go]: And why is it named `perr`?  What does the `p` stand for?

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Thank you very much. I have addressed this in the if condition.

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Thank you very much. I have addressed this in the if condition and not using this variable.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: I think only one boolean is needed...
 
```suggestion
				shouldTruncate := false

				// This is to handle a scenario in which upstream sets the TC bit, but doesn't truncate the response
				// and we get ErrBuf instead of overflow.
				if _, isDNSErr := err.(*dns.Error); isDNSErr && errors.Is(err, dns.ErrBuf) {
					shouldTruncate = true
				} else if strings.Contains(err.Error(), "overflow") {
					shouldTruncate = true
				}

				if shouldTruncate {
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Ah yes yes. Thank you. I will do better next time.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: Please make this into a function, and add unit tests for it.

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: `proto` isn't necessarily reliable to determine the actual protocol used with the upstream host.  For example, when forwarding UDP requests to upstream over DoT (dns over tls) connections, the value of proto will be "udp" here, while the actual protocol is over tcp.

Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: oh ok understood. Changed the condition here - if p.transport.transportTypeFromConn(pc) == typeUDP

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Implemented both the suggested changes.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: If the IDs don't match, we can skip checking the response truncation errors.  IOW, the ID match check should happen before the call to `shouldTruncateResponse(err)`.  ~It probably can go before the protocol check actually.~

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [reply plugin/pkg/proxy/connect.go]: >  It probably can go before the protocol check actually.

Actually - no I don't think it can.  But it should definitely go before the call to `shouldTruncateResponse`.

Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure sure, addressed this. Thank you.

**Comment by SriHarsha001** [comment]
@chrisohaver - When you get few minutes, I kindly request you to review the latest changes in this PR. Thanking you in advance.

File: plugin/pkg/proxy/connect.go
Comment (chrisohaver) [inline plugin/pkg/proxy/connect.go]: These three nested `if`s can be combined into one since instructions are only present in the inner most `if`. e.g. 

```
if ret != nil && state.Req.Id == ret.Id &&
	p.transport.transportTypeFromConn(pc) == typeUDP && shouldTruncateResponse(err) {
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/pkg/proxy/connect.go
Comment (SriHarsha001) [reply plugin/pkg/proxy/connect.go]: Sure, made the suggested changes. Thank you.

**Review by chrisohaver — APPROVED** [review]

**Pull request merged**
